src/ast/rewriter/seq_rewriter.cpp

```diff
@@ -3344,13 +3320,137 @@ expr_ref seq_rewriter::mk_antimirov_deriv_restrict(expr* e, expr* d, expr* cond)
     else if (re().is_union(d, a, b)) {
         expr_ref a1(mk_antimirov_deriv_restrict(e, a, cond), m());
         expr_ref b1(mk_antimirov_deriv_restrict(e, b, cond), m());
-        if (a1 == b1 || re().is_empty(b1) || re().is_full_seq(a1))
-            result = a1;
-        else if (re().is_empty(a1) || re().is_full_seq(b1))
-            result = b1;
-        else
-            //TODO: merge
-            result = (a1->get_id() <= b1->get_id() ? re().mk_union(a1, b1) : re().mk_union(b1, a1));
+        result = mk_regex_union_normalize(a1, b1);
+    }
+    return result;
+}
+
+expr_ref seq_rewriter::mk_regex_union_normalize(expr* r1, expr* r2) {
+    VERIFY(m_util.is_re(r1));
+    VERIFY(m_util.is_re(r2));
+    expr_ref result(m());
+    if (r1 == r2 || re().is_empty(r2) || re().is_full_seq(r1))
+        result = r1;
+    else if (re().is_empty(r1) || re().is_full_seq(r2))
+        result = r2;
+    else if (re().is_dot_plus(r1) && re().get_info(r2).min_length > 0)
+        result = r1;
+    else if (re().is_dot_plus(r2) && re().get_info(r1).min_length > 0)
+        result = r2;
+    else
+        result = mk_regex_union_merge(r1, r2);
+    return result;
+}
+// Assumes r1 and r2 are both ordered lists in right-associative form without duplicates when unions
+// Merges the two lists as a new ordered list where duplicates are removed
+// an element and its complement have the same merge-order and result in true
+expr_ref seq_rewriter::mk_regex_union_merge(expr* r1, expr* r2) {
+    expr_ref result(m());
+    expr* r1first, * r1rest, * r2first, * r2rest;
+    if (re().is_union(r1, r1first, r1rest)) {
+        if (re().is_union(r2, r2first, r2rest)) {
+            if (r1first->get_id() < r2first->get_id())
+                result = re().mk_union(r1first, mk_regex_union_merge(r1rest, r2));
```

Comment:
someone will fuzz this to stack overflow.

Also, can't you assume that r2, r1 are already in normal form?

Reply:
Here r2 and r1 are assumed to be in normal form (ordered lists essentially).
So merge makes their union keeping the result ordered. I'll change that to use a loop.